<b>Note</b>: Pada kuis kali ini kita akan menggunakan data diamonds.csv yang bisa kalian download [disini](https://drive.google.com/uc?export=download&id=1qQiM9utQUThTPq8jk8stPMh_y2zfqYOd). setelah di download, kalian pindahkan data tersebut satu level dengan direktori file notebook ini.


In [2]:
from google.colab import drive
drive.mount('/content/gdrive')

Mounted at /content/gdrive


In [3]:
import pandas as pd
df = pd.read_csv('/content/gdrive/MyDrive/PKS DIgital School/diamonds.csv')
df.drop('Unnamed: 0', axis=1, inplace=True)

In [4]:
df.head()

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31
2   0.23     Good     E     VS1   56.9   65.0    327  4.05  4.07  2.31
3   0.29  Premium     I     VS2   62.4   58.0    334  4.20  4.23  2.63
4   0.31     Good     J     SI2   63.3   58.0    335  4.34  4.35  2.75

Columns Metadata:

carat: berat diamond

cut: kualitas diamond

color: warna diamond

clarity: level kemurnian diamonds

depth: tinggi dari diamond tabel

table: lebar  dari diamond tabel

price: harga diamond ($US)

x: panjang

y: lebar

z: kedalaaman

<h2>Soal 1: Process Grouping</h2>

Jelaskan prosess yang terjadi dalam grouping data menggungkan method <b>groupby()</b> dari pandas.

jawab:
Grouping adalah suatu teknik memisahkan data berdasarkan kriteria tertentu. untuk melakukan grouping di pandas kita menggunakan suatu method yang bernama groupby(). definisi abstraknya dari grouping adalah kita melakukan mapping suatu data terhadap suatu group.

<h2>Soal 2: Grouping All Column</h2>
    
Group data diamond berdasarkan cut (kualitas), lakukan aggregasi dengan fungsi rata2.

Kemudian jawab pertanyaan berikut.
- Berapa harga rata2 dari diamond berkualitas Premium

Expected Output:

![alt text](https://drive.google.com/uc?id=1GKanuLzNv_pYBFVRBmQywxQw39uQBvSl)

In [8]:
# code here
df.groupby('cut').mean()

<ipython-input-8-fbad58b09c88>:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df.groupby('cut').mean()


carat      depth      table        price         x         y  \
cut                                                                          
Fair       1.046137  64.041677  59.053789  4358.757764  6.246894  6.182652   
Good       0.849185  62.365879  58.694639  3928.864452  5.838785  5.850744   
Ideal      0.702837  61.709401  55.951668  3457.541970  5.507451  5.520080   
Premium    0.891955  61.264673  58.746095  4584.257704  5.973887  5.944879   
Very Good  0.806381  61.818275  57.956150  3981.759891  5.740696  5.770026   

                  z  
cut                  
Fair       3.982770  
Good       3.639507  
Ideal      3.401448  
Premium    3.647124  
Very Good  3.559801

jawab:
Harga rata-rata dari diamond berkualitas Premium 0.891955

<h2>Soal 3: Grouping Some Column base on Multiple Criteria</h2>
    
Lakukan groupby pada data berdasarkan kriteria cut (kualitas) dan color (warna) secara berurutan, dengan aggregasi pada column carat, table, dan price saja. dengan fungsi agregasi rata-rata.

Kemudian jawab pertanyaan berikut:
- Berapa harga diamond yang berkriteria premium dan berwarna J
- Berapa nilai dari carat diamond yang berkriteria very good dan berwarna H

Expected Output:

Note: gambar hanya sebagaian dari keseluruhan data

![alt text](https://drive.google.com/uc?id=1XB_uHor7ZCPQUH9uQjSfd6aC6_yJnjh9)

In [11]:
# code here
df.groupby(['cut','color']).agg({'carat' : 'mean', 'table' : 'mean', 'price' : 'mean'})

carat      table        price
cut       color                                  
Fair      D      0.920123  58.969325  4291.061350
          E      0.856607  59.364732  3682.312500
          F      0.904712  59.453205  3827.003205
          G      1.023822  58.773248  4239.254777
          H      1.219175  58.696370  5135.683168
          I      1.198057  59.237143  4685.445714
          J      1.341176  58.917647  4975.655462
Good      D      0.744517  58.541541  3405.382175
          E      0.745134  58.779957  3423.644159
          F      0.775930  58.910891  3495.750275
          G      0.850896  58.471986  4123.482204
          H      0.914729  58.611111  4276.254986
          I      1.057222  58.773946  5078.532567
          J      1.099544  58.813029  4574.172638
Ideal     D      0.565766  55.965632  2629.094566
          E      0.578401  55.967461  2597.550090
          F      0.655829  55.924203  3374.939362
          G      0.700715  55.902375  3720.706388
          H      0.799525  55.965843  3889.334831
          I      0.913029  56.021357  4451.970377
          J      1.063594  56.012612  4918.186384
Premium   D      0.721547  58.718964  3631.292576
          E      0.717745  58.779461  3538.914420
          F      0.827036  58.679279  4324.890176
          G      0.841488  58.702360  4500.742134
          H      1.016449  58.792034  5216.706780
          I      1.144937  58.771849  5946.180672
          J      1.293094  58.874752  6294.591584
Very Good D      0.696424  58.041309  3470.467284
          E      0.676317  58.038875  3214.652083
          F      0.740961  57.848429  3778.820240
          G      0.766799  57.784428  3872.753806
          H      0.915948  57.903015  4535.390351
          I      1.046952  58.105150  5255.879568
          J      1.133215  58.277729  5103.513274

jawab:
Harga rata-rata dari diamond yang berkriteria premium dan berwarna J adalah 6294.591584 dan nilai rata-rata dari carat diamond yang berkriteria very good dan berwarna H adalah 0.915948

<h2>Soal 4: Group By Multiple Aggregate Base on Multiple Criteria</h2>
    
Lakukan groupby pada data berdasarkan kriteria cut(kualitas) dan color (warna) secara berurutan kemudian aggregasi objek group by nya dengan fungsi rata2 dan median.

Kemudian Jawab Pertanyaan berikut:

- Berapa nilai median dari harga diamond berkualitas good dan berwarna F

Expected Output:

Note : gambar hanya sebagaian dari keseluruhan data

![alt text](https://drive.google.com/uc?id=1fUPIse5ELrzj26ZyWBqsAVOUb6yf7UoD)

In [18]:
# code here
df.groupby(['cut','color'])[['carat','table','price']].agg(['mean','median'])

carat             table               price        
                     mean median       mean median         mean  median
cut       color                                                        
Fair      D      0.920123  0.900  58.969325   58.0  4291.061350  3730.0
          E      0.856607  0.900  59.364732   59.0  3682.312500  2956.0
          F      0.904712  0.900  59.453205   59.0  3827.003205  3035.0
          G      1.023822  0.980  58.773248   58.0  4239.254777  3057.0
          H      1.219175  1.010  58.696370   58.0  5135.683168  3816.0
          I      1.198057  1.010  59.237143   58.0  4685.445714  3246.0
          J      1.341176  1.030  58.917647   58.0  4975.655462  3302.0
Good      D      0.744517  0.700  58.541541   58.0  3405.382175  2728.5
          E      0.745134  0.700  58.779957   58.0  3423.644159  2420.0
          F      0.775930  0.710  58.910891   59.0  3495.750275  2647.0
          G      0.850896  0.900  58.471986   58.0  4123.482204  3340.0
          H      0.914729  0.900  58.611111   58.0  4276.254986  3468.5
          I      1.057222  1.000  58.773946   58.0  5078.532567  3639.5
          J      1.099544  1.020  58.813029   58.0  4574.172638  3733.0
Ideal     D      0.565766  0.500  55.965632   56.0  2629.094566  1576.0
          E      0.578401  0.500  55.967461   56.0  2597.550090  1437.0
          F      0.655829  0.530  55.924203   56.0  3374.939362  1775.0
          G      0.700715  0.540  55.902375   56.0  3720.706388  1857.5
          H      0.799525  0.700  55.965843   56.0  3889.334831  2278.0
          I      0.913029  0.740  56.021357   56.0  4451.970377  2659.0
          J      1.063594  1.030  56.012612   56.0  4918.186384  4096.0
Premium   D      0.721547  0.580  58.718964   59.0  3631.292576  2009.0
          E      0.717745  0.580  58.779461   59.0  3538.914420  1928.0
          F      0.827036  0.760  58.679279   59.0  4324.890176  2841.0
          G      0.841488  0.755  58.702360   59.0  4500.742134  2745.0
          H      1.016449  1.010  58.792034   59.0  5216.706780  4511.0
          I      1.144937  1.140  58.771849   59.0  5946.180672  4640.0
          J      1.293094  1.250  58.874752   59.0  6294.591584  5063.0
Very Good D      0.696424  0.610  58.041309   58.0  3470.467284  2310.0
          E      0.676317  0.570  58.038875   58.0  3214.652083  1989.5
          F      0.740961  0.700  57.848429   58.0  3778.820240  2471.0
          G      0.766799  0.700  57.784428   58.0  3872.753806  2437.0
          H      0.915948  0.900  57.903015   58.0  4535.390351  3734.0
          I      1.046952  1.005  58.105150   58.0  5255.879568  3888.0
          J      1.133215  1.060  58.277729   58.0  5103.513274  4113.0

jawab:
Nilai median dari harga diamond berkualitas good dan berwarna F adalah 2647.0

<h2>Soal 5: Group By With Different Aggregation for specific Column</h2>
    
Lakukan groupby pada data berdasarkan criteria cut(kualitas), kemudian aggreagate dengan fungsi max pada column price dan fungsi min pada column carat.

Kemudian Jawab Pertanyaan berikut:

- Berapa nilai max dari price untuk diamond berkualitas Ideal

Expected Output:

![alt text](https://drive.google.com/uc?id=1B_z5PBkLe4hNJJJV3JgDGP9G--llytKf)

In [20]:
# code here
df.groupby('cut').agg({'price':'max','carat':'min'})

price  carat
cut                    
Fair       18574   0.22
Good       18788   0.23
Ideal      18806   0.20
Premium    18823   0.20
Very Good  18818   0.20

jawab:
nilai max dari price untuk diamond berkualitas Ideal adalah 18806

<h2>Soal 6: Group By With Custom Function</h2>
    
Lakukan groupby pada data berdasarkan criteria cut(kualitas), kemudian lakukan aggregasi dengan fungsi standar deviasi pada column carat dan fungsi rentang nilai antara max dan min pada column price.

Kemudian jawab pertanyaan berikut:
- Berapa selisih harga tertinggi dan terendah dari diamond berkualitas Premium

Expected Output:

![alt text](https://drive.google.com/uc?id=1bl6_8KryXCxLfzpAf4YF80Qvt15a3uQ6)

In [22]:
# code here
def selisih(x):
    return x.max() - x.min()
df.groupby('cut').agg({'price':selisih,'carat':'std'})

price     carat
cut                       
Fair       18237  0.516404
Good       18461  0.454054
Ideal      18480  0.432876
Premium    18497  0.515262
Very Good  18482  0.459435

jawab:
Selisih harga tertinggi dan terendah dari diamond berkualitas Premium adalah 18497